File: ios/DicomParser.swift
import Foundation

struct DicomTag {
    let group: UInt16
    let element: UInt16
    let vr: String
    let length: UInt32
    let value: Data
    var frames: [Data]? 
    
    /// Interprets little‑endian US/SS/UL values as Swift `Int`.
    var asInt: Int? {
        guard length == 2 || length == 4 else { return nil }
        return value.withUnsafeBytes {
            length == 2
            ? Int(UInt16(littleEndian: $0.load(as: UInt16.self)))
            : Int(UInt32(littleEndian: $0.load(as: UInt32.self)))
        }
    }

    /// Returns the raw value field as `Data`.
    var asData: Data? { value }
}

class DicomParser {
    private let data: Data
    private var cursor: Int = 0

    init(fileURL: URL) throws {
        self.data = try Data(contentsOf: fileURL)
        self.cursor = 0
    }

    func parse() -> [DicomTag] {
        var tags: [DicomTag] = []

        // Skip 128-byte preamble + "DICM" (4 bytes)
        cursor = 132

        while cursor + 8 <= data.count {
            guard let tag = readTag() else { 
                break 
            }
            tags.append(tag)
        }
        return tags
    }

    private func readTag() -> DicomTag? {
        guard cursor + 8 <= data.count else { 
            return nil 
        }

        let group = readUInt16()
        let element = readUInt16()
        let vr = readString(length: 2)
        var length: UInt32 = 0
        
        // Handle special cases
        if vr == "SQ" || (group == 0x7FE0 && element == 0x0010) { // Sequence or Pixel Data
            cursor += 2 // Skip reserved bytes
            length = readUInt32()
            
            if length == 0xFFFFFFFF {                
                if vr == "SQ" {
                    // Skip sequence items until sequence delimiter
                    var sequenceData = Data()
                    while cursor + 8 <= data.count {
                        let itemGroup = readUInt16()
                        let itemElement = readUInt16()
                        
                        if itemGroup == 0xFFFE {
                            let itemLength = readUInt32()
                            if itemElement == 0xE0DD { // Sequence Delimiter
                                break
                            } else if itemElement == 0xE000 { // Item
                                if itemLength != 0xFFFFFFFF {
                                    if cursor + Int(itemLength) <= data.count {
                                        let itemData = data.subdata(in: cursor..<(cursor + Int(itemLength)))
                                        sequenceData.append(itemData)
                                        cursor += Int(itemLength)
                                    }
                                }
                            }
                        } else {
                            cursor -= 4 // Move back if not a FFFE tag
                            break
                        }
                    }
                    return DicomTag(group: group, element: element, vr: vr, length: UInt32(sequenceData.count), value: sequenceData)
                } else {
                    // Handle encapsulated pixel data
                    var frames: [Data] = []
                    
                    // Skip Basic Offset Table
                    let itemGroup = readUInt16()
                    let itemElement = readUInt16()
                    if itemGroup == 0xFFFE && itemElement == 0xE000 {
                        let offsetTableLength = readUInt32()
                        cursor += Int(offsetTableLength)
                    } else {
                        cursor -= 4
                    }
                    
                    // Read each frame
                    while cursor + 8 <= data.count {
                        let frameGroup = readUInt16()
                        let frameElement = readUInt16()
                        let frameLength = readUInt32()
                        
                        if frameGroup == 0xFFFE {
                            if frameElement == 0xE000 { // Frame Item
                                if cursor + Int(frameLength) <= data.count {
                                    let frameData = data.subdata(in: cursor..<(cursor + Int(frameLength)))
                                    if let jpegStart = frameData.firstRange(of: Data([0xFF, 0xD8])) {
                                        let jpegData = frameData.subdata(in: jpegStart.startIndex..<frameData.endIndex)
                                        frames.append(jpegData)
                                    }
                                    cursor += Int(frameLength)
                                }
                            } else if frameElement == 0xE0DD { // Sequence Delimiter
                                break
                            }
                        } else {
                            cursor -= 8
                            break
                        }
                    }
                    
                    return DicomTag(group: group, element: element, vr: vr, length: length, value: Data(), frames: frames)
                }
            }
        }
        // Handle other VRs with 2-byte or 4-byte length
        else if ["OB", "OW", "UN", "UT"].contains(vr) {
            cursor += 2 // Reserved bytes
            length = readUInt32()
        } else {
            length = UInt32(readUInt16())
        }

        guard cursor + Int(length) <= data.count else {
            return nil 
        }

        let value = data.subdata(in: cursor..<(cursor + Int(length)))
        cursor += Int(length)

        return DicomTag(group: group, element: element, vr: vr, length: length, value: value)
    }

    private func readUInt16() -> UInt16 {
        let val = data.subdata(in: cursor..<(cursor + 2)).withUnsafeBytes { $0.load(as: UInt16.self) }
        cursor += 2
        return UInt16(littleEndian: val)
    }

    private func readUInt32() -> UInt32 {
        let val = data.subdata(in: cursor..<(cursor + 4)).withUnsafeBytes { $0.load(as: UInt32.self) }
        cursor += 4
        return UInt32(littleEndian: val)
    }

    private func readString(length: Int) -> String {
        let bytes = data.subdata(in: cursor..<(cursor + length))
        cursor += length
        return String(data: bytes, encoding: .ascii)?.trimmingCharacters(in: .controlCharacters) ?? ""
    }
}

File: android/src/main/java/com/dicomviewer/DicomParser.kt
package com.dicomviewer

import android.util.Log
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

data class DicomTag(
    val group: UShort,
    val element: UShort,
    val vr: String,
    val length: UInt,
    val value: ByteArray,
    var frames: List<ByteArray>? = null
) {
    val asInt: Int?
        get() {
            if (length != 2u && length != 4u) return null
            return ByteBuffer.wrap(value)
                .order(ByteOrder.LITTLE_ENDIAN)
                .let {
                    when (length) {
                        2u -> it.short.toInt()
                        4u -> it.int
                        else -> null
                    }
                }
        }

    val asData: ByteArray?
        get() = value
}

class DicomParser(private val file: File) {
    private var data: ByteArray = file.readBytes()
    private var cursor: Int = 0

    fun parse(): List<DicomTag> {
        Log.d("DicomParser", "Starting to parse DICOM file: ${file.name}")
        val tags = mutableListOf<DicomTag>()

        // Skip 128-byte preamble + "DICM" (4 bytes)
        cursor = 132
        Log.d("DicomParser", "Skipped preamble, starting at position: $cursor")

        while (cursor + 8 <= data.size) {
            val tag = readTag()
            if (tag != null) {
                Log.d("DicomParser", "Read tag: Group=${tag.group.toString(16)}, Element=${tag.element.toString(16)}, VR=${tag.vr}, Length=${tag.length}")
                tags.add(tag)
            } else {
                Log.d("DicomParser", "Failed to read tag at position: $cursor")
                break
            }
        }

        Log.d("DicomParser", "Finished parsing, found ${tags.size} tags")
        return tags
    }

    private fun readTag(): DicomTag? {
        if (cursor + 8 > data.size) {
            Log.d("DicomParser", "Not enough data left to read tag header")
            return null
        }

        val group = readUShort()
        val element = readUShort()
        val vr = readString(2)
        var length: UInt = 0u

        Log.d("DicomParser", "Reading tag: Group=${group.toString(16)}, Element=${element.toString(16)}, VR=$vr")

        // Handle special cases
        if (vr == "SQ" || (group == 0x7FE0.toUShort() && element == 0x0010.toUShort())) { // Sequence or Pixel Data
            cursor += 2 // Skip reserved bytes
            length = readUInt()
            
            if (length == 0xFFFFFFFFu) {
                Log.d("DicomParser", "Found undefined length sequence/pixel data")
                
                if (vr == "SQ") {
                    Log.d("DicomParser", "Processing sequence items")
                    // Skip sequence items until sequence delimiter
                    val sequenceData = mutableListOf<Byte>()
                    while (cursor + 8 <= data.size) {
                        val itemGroup = readUShort()
                        val itemElement = readUShort()
                        
                        if (itemGroup == 0xFFFEu.toUShort()) {
                            val itemLength = readUInt()
                            if (itemElement == 0xE0DDu.toUShort()) { // Sequence Delimiter
                                Log.d("DicomParser", "Found sequence delimiter")
                                break
                            } else if (itemElement == 0xE000u.toUShort()) { // Item
                                Log.d("DicomParser", "Processing sequence item of length: $itemLength")
                                if (itemLength != 0xFFFFFFFFu) {
                                    if (cursor + itemLength.toInt() <= data.size) {
                                        sequenceData.addAll(data.slice(cursor until cursor + itemLength.toInt()))
                                        cursor += itemLength.toInt()
                                    }
                                }
                            }
                        } else {
                            cursor -= 4 // Move back if not a FFFE tag
                            break
                        }
                    }
                    return DicomTag(group, element, vr, sequenceData.size.toUInt(), sequenceData.toByteArray())
                } else {
                    Log.d("DicomParser", "Processing encapsulated pixel data")
                    // Handle encapsulated pixel data
                    val frames = mutableListOf<ByteArray>()
                    
                    // Skip Basic Offset Table
                    val itemGroup = readUShort()
                    val itemElement = readUShort()
                    if (itemGroup == 0xFFFEu.toUShort() && itemElement == 0xE000u.toUShort()) {
                        val offsetTableLength = readUInt()
                        Log.d("DicomParser", "Skipping offset table of length: $offsetTableLength")
                        cursor += offsetTableLength.toInt()
                    } else {
                        cursor -= 4
                    }
                    
                    // Read each frame
                    while (cursor + 8 <= data.size) {
                        val frameGroup = readUShort()
                        val frameElement = readUShort()
                        val frameLength = readUInt()
                        
                        if (frameGroup == 0xFFFEu.toUShort()) {
                            if (frameElement == 0xE000u.toUShort()) { // Frame Item
                                Log.d("DicomParser", "Reading frame of length: $frameLength")
                                if (cursor + frameLength.toInt() <= data.size) {
                                    val frameData = data.slice(cursor until cursor + frameLength.toInt()).toByteArray()
                                    // Find JPEG start marker (FF D8)
                                    val jpegStartIndex = frameData.findJpegStart()
                                    if (jpegStartIndex != -1) {
                                        Log.d("DicomParser", "Found JPEG start marker at offset: $jpegStartIndex")
                                        frames.add(frameData.slice(jpegStartIndex until frameData.size).toByteArray())
                                    }
                                    cursor += frameLength.toInt()
                                }
                            } else if (frameElement == 0xE0DDu.toUShort()) { // Sequence Delimiter
                                Log.d("DicomParser", "Found pixel data sequence delimiter")
                                break
                            }
                        } else {
                            cursor -= 8
                            break
                        }
                    }
                    
                    return DicomTag(group, element, vr, length, ByteArray(0), frames)
                }
            }
        }
        // Handle other VRs with 2-byte or 4-byte length
        else if (listOf("OB", "OW", "UN", "UT").contains(vr)) {
            cursor += 2 // Reserved bytes
            length = readUInt()
            Log.d("DicomParser", "Reading ${vr} value of length: $length")
        } else {
            length = readUShort().toUInt()
            Log.d("DicomParser", "Reading standard value of length: $length")
        }

        if (cursor + length.toInt() > data.size) {
            Log.d("DicomParser", "Not enough data left to read value of length: $length")
            return null
        }

        val value = data.slice(cursor until cursor + length.toInt()).toByteArray()
        cursor += length.toInt()

        return DicomTag(group, element, vr, length, value)
    }

    private fun readUShort(): UShort {
        val bytes = data.slice(cursor until cursor + 2).toByteArray()
        cursor += 2
        return ByteBuffer.wrap(bytes)
            .order(ByteOrder.LITTLE_ENDIAN)
            .short
            .toUShort()
    }

    private fun readUInt(): UInt {
        val bytes = data.slice(cursor until cursor + 4).toByteArray()
        cursor += 4
        return ByteBuffer.wrap(bytes)
            .order(ByteOrder.LITTLE_ENDIAN)
            .int
            .toUInt()
    }

    private fun readString(length: Int): String {
        val bytes = data.slice(cursor until cursor + length).toByteArray()
        cursor += length
        return String(bytes).trim { it <= ' ' }
    }
}

private fun ByteArray.findJpegStart(): Int {
    for (i in 0..size - 2) {
        if (this[i] == 0xFF.toByte() && this[i + 1] == 0xD8.toByte()) {
            return i
        }
    }
    return -1
} 